habitat-lab/habitat/tasks/rearrange/utils.py

```diff
@@ -785,3 +606,197 @@ def get_camera_lookat_relative_to_vertial_line(
+
+                    # translation to global 3D points at navmesh height
+                    offsets_3d = [
+                        np.array(
+                            [
+                                transformed_offset_2d[0],
+                                0,
+                                transformed_offset_2d[1],
+                            ]
+                        )
+                        + batch_sample[0]
+                        for transformed_offset_2d in transformed_offsets_2d
+                    ]
+
+                    if data_out is not None:
+                        data_out["offsets_3d"] = offsets_3d
+
+                    # check for offset navigability
+                    is_collision = False
+                    for offset_point in offsets_3d:
+                        if not (
+                            sim.pathfinder.is_navigable(offset_point)
+                            and (
+                                island_id == -1
+                                or sim.pathfinder.get_island(offset_point)
+                                == island_id
+                            )
+                        ):
+                            is_collision = True
+                            break
+
+                    # if this sample is invalid, try the next
+                    if is_collision:
+                        continue
+
+                if agent_embodiement is not None:
+                    # contact testing with collision shapes
+                    start_position = agent_embodiement.base_pos
+                    start_rotation = agent_embodiement.base_rot
+
+                    agent_embodiement.base_pos = batch_sample[0]
+                    agent_embodiement.base_rot = desired_angle
+
+                    details = None
+                    # Make sure the robot is not colliding with anything in this state.
+                    if sim.__class__.__name__ == "RearrangeSim":
+                        _, details = rearrange_collision(
+                            sim,
+                            False,
+                            ignore_base=False,
+                        )
+                    else:
+                        _, details = general_sim_collision(
+                            sim, agent_embodiement
+                        )
+
+                    # reset agent state
+                    agent_embodiement.base_pos = start_position
+                    agent_embodiement.base_rot = start_rotation
+
+                    # Only care about collisions between the robot and scene.
+                    is_feasible_state = details.robot_scene_colls == 0
+                    if not is_feasible_state:
+                        continue
+
+                # if we made it here, all tests passed and we found a valid placement state
+                return batch_sample[0], desired_angle, True
+
+    # unable to find a valid navmesh point within constraints
+    return None, None, False
```

Comment:
at least we can return something to set the agent pose after best effect of searching?